File: backend-python/src/prompt/_internal/entities.py
from __future__ import annotations

from datetime import datetime

from sqlalchemy import Boolean, Column, DateTime, Integer, String, Text, func
from sqlalchemy.orm import relationship
from sqlmodel import Field, Relationship, SQLModel


class PromptCategory(SQLModel, table=True):
    __tablename__ = "prompt_categories"
    __allow_unmapped__ = True

    id: int | None = Field(default=None, primary_key=True)
    name: str = Field(sa_column=Column(String(255), unique=True, nullable=False))

    # Relationships
    subcategories: list[PromptSubCategory] = Relationship(
        sa_relationship=relationship("PromptSubCategory", back_populates="prompt_category")
    )

    created_at: datetime | None = Field(
        default=None,
        sa_column=Column(DateTime(timezone=True), nullable=False, server_default=func.now()),
    )
    updated_at: datetime | None = Field(
        default=None,
        sa_column=Column(
            DateTime(timezone=True),
            nullable=False,
            server_default=func.now(),
            onupdate=func.now(),
        ),
    )

    def __repr__(self) -> str:  # pragma: no cover - debug representation only
        return f"PromptCategory(id={self.id!r}, name={self.name!r})"


class PromptSubCategory(SQLModel, table=True):
    __tablename__ = "prompt_subcategories"
    __allow_unmapped__ = True

    id: int | None = Field(default=None, primary_key=True)

    prompt_category_id: int = Field(foreign_key="prompt_categories.id", nullable=False)
    prompt_category: PromptCategory | None = Relationship(
        sa_relationship=relationship("PromptCategory", back_populates="subcategories")
    )

    name: str = Field(sa_column=Column(String(255), nullable=False))
    description: str | None = Field(default=None, sa_column=Column(Text, nullable=True))

    created_at: datetime | None = Field(
        default=None,
        sa_column=Column(DateTime(timezone=True), nullable=False, server_default=func.now()),
    )
    updated_at: datetime | None = Field(
        default=None,
        sa_column=Column(
            DateTime(timezone=True),
            nullable=False,
            server_default=func.now(),
            onupdate=func.now(),
        ),
    )

    def __repr__(self) -> str:  # pragma: no cover - debug representation only
        return f"PromptSubCategory(id={self.id!r}, name={self.name!r}, category_id={self.prompt_category_id!r})"


class PromptSlotType(SQLModel, table=True):
    __tablename__ = "prompt_slot_types"
    __allow_unmapped__ = True

    id: int | None = Field(default=None, primary_key=True)
    name: str = Field(sa_column=Column(String(255), unique=True, nullable=False))
    position: int = Field(sa_column=Column(Integer, unique=True, nullable=False))

    created_at: datetime | None = Field(
        default=None,
        sa_column=Column(DateTime(timezone=True), nullable=False, server_default=func.now()),
    )
    updated_at: datetime | None = Field(
        default=None,
        sa_column=Column(
            DateTime(timezone=True),
            nullable=False,
            server_default=func.now(),
            onupdate=func.now(),
        ),
    )

    def __repr__(self) -> str:  # pragma: no cover - debug representation only
        return f"PromptSlotType(id={self.id!r}, name={self.name!r}, position={self.position!r})"


class PromptSlotVariant(SQLModel, table=True):
    __tablename__ = "prompt_slot_variants"
    __allow_unmapped__ = True

    id: int | None = Field(default=None, primary_key=True)

    prompt_slot_type_id: int = Field(foreign_key="prompt_slot_types.id", nullable=False)
    prompt_slot_type: PromptSlotType | None = Relationship(sa_relationship=relationship("PromptSlotType"))

    name: str = Field(sa_column=Column(String(255), unique=True, nullable=False))
    prompt: str | None = Field(default=None, sa_column=Column(Text, nullable=True))
    description: str | None = Field(default=None, sa_column=Column(Text, nullable=True))
    example_image_filename: str | None = Field(default=None, sa_column=Column(String(500), nullable=True))

    # Backref to mapping table
    prompt_slot_variant_mappings: list[PromptSlotVariantMapping] = Relationship(
        sa_relationship=relationship("PromptSlotVariantMapping", back_populates="prompt_slot_variant")
    )

    created_at: datetime | None = Field(
        default=None,
        sa_column=Column(DateTime(timezone=True), nullable=False, server_default=func.now()),
    )
    updated_at: datetime | None = Field(
        default=None,
        sa_column=Column(
            DateTime(timezone=True),
            nullable=False,
            server_default=func.now(),
            onupdate=func.now(),
        ),
    )

    def __repr__(self) -> str:  # pragma: no cover - debug representation only
        return f"PromptSlotVariant(id={self.id!r}, name={self.name!r})"


class Prompt(SQLModel, table=True):
    __tablename__ = "prompts"
    __allow_unmapped__ = True

    id: int | None = Field(default=None, primary_key=True)

    title: str = Field(sa_column=Column(String(500), nullable=False))
    prompt_text: str | None = Field(default=None, sa_column=Column(Text, nullable=True))

    category_id: int | None = Field(default=None, foreign_key="prompt_categories.id")
    category: PromptCategory | None = Relationship(sa_relationship=relationship("PromptCategory", lazy="joined"))

    subcategory_id: int | None = Field(default=None, foreign_key="prompt_subcategories.id")
    subcategory: PromptSubCategory | None = Relationship(
        sa_relationship=relationship("PromptSubCategory", lazy="joined")
    )

    active: bool = Field(default=True, sa_column=Column(Boolean, nullable=False, server_default="true"))
    example_image_filename: str | None = Field(default=None, sa_column=Column(String(500), nullable=True))

    # Mapping relations
    prompt_slot_variant_mappings: list[PromptSlotVariantMapping] = Relationship(
        sa_relationship=relationship("PromptSlotVariantMapping", back_populates="prompt")
    )

    created_at: datetime | None = Field(
        default=None,
        sa_column=Column(DateTime(timezone=True), nullable=False, server_default=func.now()),
    )
    updated_at: datetime | None = Field(
        default=None,
        sa_column=Column(
            DateTime(timezone=True),
            nullable=False,
            server_default=func.now(),
            onupdate=func.now(),
        ),
    )

    def __repr__(self) -> str:  # pragma: no cover - debug representation only
        return f"Prompt(id={self.id!r}, title={self.title!r}, active={self.active!r})"


class PromptSlotVariantMapping(SQLModel, table=True):
    """Association entity between Prompt and PromptSlotVariant.

    Composite PK: (prompt_id, slot_id). Stores creation timestamp for mapping.
    """

    __tablename__ = "prompt_slot_variant_mappings"
    __allow_unmapped__ = True

    prompt_id: int | None = Field(default=None, foreign_key="prompts.id", primary_key=True)
    slot_id: int | None = Field(default=None, foreign_key="prompt_slot_variants.id", primary_key=True)

    # Relationships
    prompt: Prompt | None = Relationship(
        sa_relationship=relationship("Prompt", back_populates="prompt_slot_variant_mappings")
    )
    prompt_slot_variant: PromptSlotVariant | None = Relationship(
        sa_relationship=relationship("PromptSlotVariant", back_populates="prompt_slot_variant_mappings")
    )

    created_at: datetime | None = Field(
        default=None,
        sa_column=Column(DateTime(timezone=True), nullable=False, server_default=func.now()),
    )

    def __repr__(self) -> str:  # pragma: no cover - debug representation only
        return f"PromptSlotVariantMapping(prompt_id={self.prompt_id!r}, slot_id={self.slot_id!r})"

File: backend/internal/prompt/entities.go
package prompt

import (
	"time"
	"voenix/backend/internal/article"
)

// GORM models mirroring Kotlin/Python prompt entities.

type PromptCategory struct {
	ID        int       `gorm:"primaryKey" json:"id"`
	Name      string    `gorm:"size:255;uniqueIndex;not null" json:"name"`
	CreatedAt time.Time `json:"createdAt"`
	UpdatedAt time.Time `json:"updatedAt"`
}

func (PromptCategory) TableName() string { return "prompt_categories" }

type PromptSubCategory struct {
	ID               int             `gorm:"primaryKey" json:"id"`
	PromptCategoryID int             `gorm:"not null" json:"promptCategoryId"`
	PromptCategory   *PromptCategory `gorm:"foreignKey:PromptCategoryID" json:"promptCategory,omitempty"`
	Name             string          `gorm:"size:255;not null" json:"name"`
	Description      *string         `gorm:"type:text" json:"description"`
	CreatedAt        time.Time       `json:"createdAt"`
	UpdatedAt        time.Time       `json:"updatedAt"`
}

func (PromptSubCategory) TableName() string { return "prompt_subcategories" }

type PromptSlotType struct {
	ID        int       `gorm:"primaryKey" json:"id"`
	Name      string    `gorm:"size:255;uniqueIndex;not null" json:"name"`
	Position  int       `gorm:"uniqueIndex;not null" json:"position"`
	CreatedAt time.Time `json:"createdAt"`
	UpdatedAt time.Time `json:"updatedAt"`
}

func (PromptSlotType) TableName() string { return "prompt_slot_types" }

type PromptSlotVariant struct {
	ID                   int             `gorm:"primaryKey" json:"id"`
	PromptSlotTypeID     int             `gorm:"column:slot_type_id;not null" json:"promptSlotTypeId"`
	PromptSlotType       *PromptSlotType `gorm:"foreignKey:PromptSlotTypeID;references:ID" json:"promptSlotType,omitempty"`
	Name                 string          `gorm:"size:255;uniqueIndex;not null" json:"name"`
	Prompt               *string         `gorm:"type:text" json:"prompt"`
	Description          *string         `gorm:"type:text" json:"description"`
	ExampleImageFilename *string         `gorm:"size:500" json:"exampleImageFilename"`
	CreatedAt            time.Time       `json:"createdAt"`
	UpdatedAt            time.Time       `json:"updatedAt"`
}

func (PromptSlotVariant) TableName() string { return "prompt_slot_variants" }

// Association table with composite PK: (prompt_id, slot_id)
type PromptSlotVariantMapping struct {
	PromptID          int                `gorm:"primaryKey;column:prompt_id" json:"promptId"`
	SlotID            int                `gorm:"primaryKey;column:slot_id" json:"slotId"`
	Prompt            *Prompt            `gorm:"foreignKey:PromptID;references:ID" json:"prompt,omitempty"`
	PromptSlotVariant *PromptSlotVariant `gorm:"foreignKey:SlotID;references:ID" json:"promptSlotVariant,omitempty"`
	CreatedAt         time.Time          `json:"createdAt"`
}

func (PromptSlotVariantMapping) TableName() string { return "prompt_slot_variant_mappings" }

type Prompt struct {
	ID                   int                      `gorm:"primaryKey" json:"id"`
	Title                string                   `gorm:"size:500;not null" json:"title"`
	PromptText           *string                  `gorm:"type:text" json:"promptText"`
	CategoryID           *int                     `gorm:"column:category_id" json:"categoryId"`
	Category             *PromptCategory          `gorm:"foreignKey:CategoryID;references:ID" json:"category,omitempty"`
	SubcategoryID        *int                     `gorm:"column:subcategory_id" json:"subcategoryId"`
	Subcategory          *PromptSubCategory       `gorm:"foreignKey:SubcategoryID;references:ID" json:"subcategory,omitempty"`
	PriceID              *int                     `gorm:"column:price_id" json:"priceId"`
	Price                *article.CostCalculation `gorm:"foreignKey:PriceID;references:ID" json:"-"`
	Active               bool                     `gorm:"not null;default:true" json:"active"`
	ExampleImageFilename *string                  `gorm:"size:500" json:"exampleImageFilename"`
	// mappings relation (manually loaded)
	PromptSlotVariantMappings []PromptSlotVariantMapping `gorm:"foreignKey:PromptID;references:ID" json:"promptSlotVariantMappings,omitempty"`
	CreatedAt                 time.Time                  `json:"createdAt"`
	UpdatedAt                 time.Time                  `json:"updatedAt"`
}

func (Prompt) TableName() string { return "prompts" }
